### PR TITLE
[CU-86b58fm3k] Added support for workflow labels when creating or upd…

### PR DESCRIPTION
**Pull request by tianning-zhang**
…ating a workflow.

### CODE REVIEW

**Comment by platform-automation-dnastack** [comment]
Task linked: [CU-86b58fm3k Update CLI to allow specifying labels on workflows](https://app.clickup.com/t/86b58fm3k)

**Review by patmagee — CHANGES_REQUESTED** [review]
## Code Review: PR #90 - Workflow Labels Support

### **Overall Assessment: ⚠️ NEEDS IMPROVEMENTS**

The PR adds workflow labels support but has several implementation issues that should be addressed before merging.

### **Positive Aspects ✅**

1. **Consistent Pattern**: Follows existing patterns for optional fields like `authors` and `description`
2. **Complete Feature**: Covers both create and update operations
3. **Test Coverage**: Updates existing tests to include label validation
4. **API Consistency**: Maintains consistent API design with other workflow fields

### **Issues and Concerns ❌**

#### **1. CLI Interface Design Issues**

```python
# In create_workflow function:
labels=labels.split(',') if labels else None
```

**Problem**: The CLI accepts labels as a single comma-separated string, but this is inconsistent with the `authors` field which also handles lists.

**Recommendation**: Use the same pattern as `authors` for consistency.

#### **2. HTTP Client Implementation Issue**

```python
# In client.py:
'labels': (None, ','.join(workflow_create_request.labels)) if workflow_create_request.labels else None
```

**Problem**: This converts a list back to a comma-separated string, which seems like unnecessary round-trip serialization.

**Recommendation**: Check if the API expects a list or comma-separated string and adjust accordingly.

#### **3. Patch Function Logic Error**

```python
def _get_labels_patch(labels: str) -> Union[JsonPatch, None]:
    if labels == "":
        return JsonPatch(path="/labels", op="remove")
    elif labels:
        return JsonPatch(path="/labels", op="replace", value=labels.split(","))
    return None
```

**Problem**: This doesn't handle `None` input properly and the empty string check should be more robust.

**Recommendation**:
```python
def _get_labels_patch(labels: Optional[str]) -> Union[JsonPatch, None]:
    if labels is None:
        return None
    if labels.strip() == "":
        return JsonPatch(path="/labels", op="remove")
    return JsonPatch(path="/labels", op="replace", value=[l.strip() for l in labels.split(",") if l.strip()])
```

#### **4. Test Quality Issues**

```python
# In test:
self.assertEqual(edited_workflow.authors, ['foo', 'bar']),  # Extra comma\!
self.assertEqual(edited_workflow.labels, ['alpha', 'beta'])
```

**Problem**: Syntax error with trailing comma in assertion.

**Recommendation**: Remove the trailing comma and use proper assertion formatting.

#### **5. Missing Edge Case Handling**

The implementation doesn't handle:
- Empty labels in comma-separated input (`"alpha,,beta"`)
- Whitespace handling around labels
- Label validation (length, characters)
- Duplicate labels

### **Specific Code Fixes Needed**

#### **Fix 1: Improve patch function**
```python
def _get_labels_patch(labels: Optional[str]) -> Union[JsonPatch, None]:
    if labels is None:
        return None
    if labels.strip() == "":
        return JsonPatch(path="/labels", op="remove")
    
    # Clean and validate labels
    cleaned_labels = [l.strip() for l in labels.split(",") if l.strip()]
    if not cleaned_labels:
        return JsonPatch(path="/labels", op="remove")
    
    return JsonPatch(path="/labels", op="replace", value=cleaned_labels)
```

#### **Fix 2: Fix test syntax**
```python
self.assertEqual(edited_workflow.authors, ['foo', 'bar'])  # Remove trailing comma
self.assertEqual(edited_workflow.labels, ['alpha', 'beta'])
```

#### **Fix 3: Add input validation**
```python
# In create_workflow function:
if labels:
    label_list = [l.strip() for l in labels.split(',') if l.strip()]
    if not label_list:
        label_list = None
else:
    label_list = None
```

### **Additional Recommendations**

#### **1. Enhanced Help Text**
```python
ArgumentSpec(
    name='labels',
    arg_names=['--labels'],
    help='Comma-separated list of labels for the workflow (e.g., "tag1,tag2,tag3")',
),
```

#### **2. Add More Test Cases**
```python
def test_workflow_labels_edge_cases():
    # Test empty string handling
    # Test whitespace handling
    # Test empty labels in comma-separated string
    # Test label removal
```

#### **3. Follow-up Documentation**
Create a follow-up ticket in dnastack-product-docs to document the new workflow labels feature, including:
- CLI usage examples
- API parameter documentation
- Best practices for label naming conventions

### **Overall Recommendation**

**REQUEST CHANGES** - The PR implements the core functionality correctly but needs refinement in:

1. Input validation and error handling
2. Consistent API design patterns
3. Fix syntax errors in tests
4. Enhanced edge case handling
5. Better documentation

The feature is valuable and follows existing patterns well, but the implementation details need polish before merging.

**Review by patmagee — CHANGES_REQUESTED** [review]
## 🚀 Ultra-Fast Parallel Review Analysis (8 agents, ~6 seconds)

**Overall Assessment: CHANGES REQUESTED** 
Risk Level: Medium | Overall Score: 75/100

### 🔴 Critical Issues Requiring Immediate Attention

1. **Runtime Error** - Missing import for `_get_labels_patch` in `commands.py:353`
2. **Syntax Error** - Trailing comma in test file will prevent test execution  
3. **Input Validation Bypass** - Labels parameter processes unsanitized input
4. **Missing Unit Tests** - Core utility functions lack proper test coverage

### 🟡 Security Concerns (Agent Score: 72/100)

- **Input Validation Gap**: `string.split(',')` without sanitization allows malicious input
- **JSON Patch Injection Risk**: Special characters in labels could cause injection  
- **No Format Validation**: Missing constraints on label format and length

### 🟡 Testing Gaps (Agent Score: 60/100)

- No unit tests for `_get_labels_patch` utility function
- Missing edge case coverage for comma-separated parsing
- No validation testing for malformed input scenarios

### ✅ Positive Aspects

- **Excellent Performance**: 95/100 score - minimal overhead, efficient implementation
- **Consistent Architecture**: Follows established patterns for similar functionality
- **Backward Compatible**: No breaking changes to existing API
- **Good Business Logic**: Correct implementation of label management

### 🔧 Required Changes

1. **Fix Critical Issues**:
   ```python
   # Add to imports in commands.py
   from dnastack.cli.workbench.utils import _get_labels_patch
   
   # Remove trailing comma in test file
   ```

2. **Add Input Validation**:
   ```python
   def _get_labels_patch(labels: Optional[str]) -> List[JsonPatchOperation]:
       if labels is None:
           return []
       if not labels.strip():
           return [JsonPatchOperation(op="remove", path="/labels")]
       
       # Add validation and sanitization
       clean_labels = [label.strip() for label in labels.split(',') if label.strip()]
       return [JsonPatchOperation(op="add", path="/labels", value=clean_labels)]
   ```

3. **Add Unit Tests**:
   - Test `_get_labels_patch` with various inputs
   - Test edge cases: empty strings, malformed input
   - Test comma-separated parsing logic

### 📊 Agent Performance Metrics
- **Code Quality**: 75/100 (good patterns, needs validation)
- **Security**: 72/100 (input validation gaps) 
- **Performance**: 95/100 (excellent efficiency)
- **Testing**: 60/100 (missing unit tests)
- **Documentation**: 70/100 (needs examples)
- **Dependencies**: 75/100 (no changes)
- **Code Style**: 75/100 (import missing)
- **Business Logic**: 80/100 (correct logic)

Great work on the feature implementation\! The architecture is sound and follows established patterns. Please address the critical issues and consider the security/testing improvements before merging.

🤖 Generated with [Claude Code](https://claude.ai/code) - Ultra-Fast Parallel PR Review

**Review by mpanik — APPROVED** [review]
LGTM

File: tests/unit/cli/workbench/test_workbench_utils.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+import unittest
+
+class TestWorkflowUtils(unittest.TestCase):
+    """Unit tests for workflow utilities"""
+
+    def test_get_labels_patch_none_input(self):
+        """Test that None input returns None"""
+        from dnastack.cli.commands.workbench.workflows.utils import _get_labels_patch
```
Comment (philippasdev) [inline tests/unit/cli/workbench/test_workbench_utils.py:8]: Why are in each test the method imported? Can't this be done once?

File: tests/unit/cli/workbench/test_workbench_utils.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        result = _get_labels_patch(None)
+        self.assertIsNone(result)
+
+    def test_get_labels_patch_empty_string(self):
+        """Test that empty string returns remove operation"""
+        from dnastack.cli.commands.workbench.workflows.utils import _get_labels_patch
+        result = _get_labels_patch("")
+        self.assertIsNotNone(result)
```
Comment (philippasdev) [inline tests/unit/cli/workbench/test_workbench_utils.py:16]: Assertpy has been added that is the pyhton version of assertJ. You could use that instead as these matchers.

File: tests/unit/cli/workbench/test_workbench_utils.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+class TestWorkflowUtils(unittest.TestCase):
+    """Unit tests for workflow utilities"""
+
+    def test_get_labels_patch_none_input(self):
+        """Test that None input returns None"""
+        from dnastack.cli.commands.workbench.workflows.utils import _get_labels_patch
+        result = _get_labels_patch(None)
+        self.assertIsNone(result)
+
+    def test_get_labels_patch_empty_string(self):
```
Comment (philippasdev) [inline tests/unit/cli/workbench/test_workbench_utils.py:12]: the description under the test name is more descriptive than the test name. Do you think it would make sense to make the test name more descriptive and remove the comment?

File: tests/cli/test_workbench.py [hunk 4]
```diff
@@ -922,11 +922,12 @@ def test_create_workflow():
 
         test_create_workflow()
 
-        def test_name_and_description():
+        def test_name_description_and_label():
```
Comment (philippasdev) [inline tests/cli/test_workbench.py:925]: Is this testname intended? I do not understand what we are testing here

**Review by philippasdev — COMMENTED** [review]
I wonder if we want to consider rolling out CDC tests at some point which would remove the need for the tight coupling of the services to the client. Also aren't the tests duplicated here?(https://docs.google.com/document/d/1hs6lVP71M8VTGL1UCyqU8b9ClZ06hWVxrON35LnH_rk/edit?tab=t.0#heading=h.p3rni8tda4hj)

**Review by philippasdev — CHANGES_REQUESTED** [review]
Please merge atleast the test_explorer test classes. Thanks :)

**Review by philippasdev — APPROVED** [review]
